### ISSUE TITLE
Milestone: Multi-user support

### ISSUE THREAD

**Issue by meichthys** [TriliumNext/Trilium#4956]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%24832%20Funded-%2300A156.svg" />](https://oss.issuehunt.io/r/TriliumNext/Trilium/issues/4956)

Many feature requests and issues exist for multiple user support. Some of the most relevant issues in the original Trilium repo are:
(https://github.com/zadam/trilium/issues/3546, https://github.com/zadam/trilium/discussions/1375, https://github.com/zadam/trilium/issues/2679, https://github.com/zadam/trilium/issues/2143, https://github.com/zadam/trilium/issues/2109, https://github.com/zadam/trilium/issues/1527, https://github.com/zadam/trilium/issues/1436, https://github.com/zadam/trilium/issues/964, https://github.com/zadam/trilium/issues/600, https://github.com/zadam/trilium/issues/450, https://github.com/zadam/trilium/issues/169)

Instead of migrating each of the issues, we will link to them here, and if desired, Zadam can close the issues on his repo and link to this issue.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $832.00)

- [<img src='https://avatars.githubusercontent.com/u/245382?v=4' alt='eitch' width=24 height=24> eitch](https://oss.issuehunt.io/u/eitch) ($52.00)
- [<img src='https://avatars.githubusercontent.com/u/4681318?v=4' alt='deajan' width=24 height=24> deajan](https://oss.issuehunt.io/u/deajan) ($500.00)
- [<img src='https://avatars.githubusercontent.com/u/1061866?v=4' alt='metrax' width=24 height=24> metrax](https://oss.issuehunt.io/u/metrax) ($50.00)
- [<img src='https://avatars.githubusercontent.com/u/42046082?v=4' alt='werererer' width=24 height=24> werererer](https://oss.issuehunt.io/u/werererer) ($20.00)
- [<img src='https://avatars.githubusercontent.com/u/5337145?v=4' alt='sirhaug' width=24 height=24> sirhaug](https://oss.issuehunt.io/u/sirhaug) ($200.00)
- [<img src='https://avatars.githubusercontent.com/u/3864731?v=4' alt='josegabrielhc' width=24 height=24> josegabrielhc](https://oss.issuehunt.io/u/josegabrielhc) ($10.00)

### Submitted pull Requests
- [#7441 feat: add multi-user support (issue #4956)](https://oss.issuehunt.io/r/TriliumNext/Trilium/pull/7441)
---

### Tips

- Checkout the [Issuehunt explorer](https://oss.issuehunt.io/r/TriliumNext/Trilium/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://oss.issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by deajan** [comment]
> it's not going to happen soon unless there is more interest in this feature, in terms of man-power and maybe even a financial incentive.
> 
> It's a really big amount of work, we're talking hundreds of hours.

Hmmm... What financial incentive would you need ? I could sponsor this a bit, and I'm sure I am not the only one.

**Comment by eliandoran** [comment]
@deajan , I'm thinking of more like a bounty where everyone can donate rather than specifying an amount on my side.
Either way, working on such a big feature can realistically only be done if I take some unpaid time off work for a few weeks.

**Comment by deajan** [comment]
Yep, I thought so. I could go for like 500-600$, but that's far from what is requested.
How about you setup that bounty ?

**Comment by eliandoran** [comment]
@deajan , @meichthys , there you go: https://oss.issuehunt.io/r/TriliumNext/Trilium/issues/4956

**Comment by meichthys** [comment]
Not sure what's going on, but I get this when submitting a bounty payment via normal mastercard:

<img width="272" height="180" alt="Image" src="https://github.com/user-attachments/assets/bd1eaf52-d398-4318-a0c6-bcd2e0feabec" />

https://support.stripe.com/questions/3ds-mandate-in-japan?locale=en-GB

If anyone else sees this, we can report it upstream.

**Comment by eliandoran** [comment]
> Not sure what's going on, but I get this when submitting a bounty payment via normal mastercard:
> 
> <img width="272" height="180" alt="Image" src="https://github.com/user-attachments/assets/bd1eaf52-d398-4318-a0c6-bcd2e0feabec" />
> 
> https://support.stripe.com/questions/3ds-mandate-in-japan?locale=en-GB
> 
> If anyone else sees this, we can report it upstream.

That's really weird, to be honest. If it causes many issues like this, we might as well switch to something else. I'm open to suggestions.

**Comment by eitch** [comment]
I just stumbled upon this application and i am blown away at its feature set. You have desktop apps as well and that is awesome. Would love the multi-user feature!

**Comment by deajan** [comment]
Maybe we need to make this bounty more public, like put it in the next release announcement or so, in order to gain sufficient traction.

**Comment by eliandoran** [comment]
@deajan , good point. 👍🏻

**Comment by Ayushkiller** [comment]
HI,i will be trying to implement this in parallel and hopefully doing it. https://github.com/Ayushkiller/Trilium I will be constantly pushing changes in stages .Currently i am actively updating some legacies and extending support for the multi user system.Will be releasing a feat 1 for everyone to see on 16 of this month hopefully with some features

**Comment by eliandoran** [comment]
@Ayushkiller , sounds good. However if you wish to receive the bounty we'd have to agree on some rules first. It's best to first provide a technical document with how you plan to implement it. And code quality is also an important factor, for this core functionality it's best not to use LLM code generation.

**Comment by deajan** [comment]
I'll add a bit about the functional point of view.
The basics should be user and group management IMO. The permissions could be as simple as CRUD.
The complicated point would be permission inheritance from nodes to sub nodes, but would be necessary in order to only give permissions on one node without the need to actually give permission on each subnode.
If "dynamic" inheritance is hard to achieve, a for loop applying parent permissions to sub nodes would also work, but would be less optimal for clones where inheritance would be overwritten if other permissions are given.

Anyway, I hope we'll get the traction to get something. This would allow Trilium to enter the enterprise environments.

**Comment by Ayushkiller** [comment]
@eliandoran  I mean i was making it regardless.I would also give the technical doc for it soon and was working on it for a while i just found about this one yesterday when skimming through issues.Still deciding on inheritances for subnodes. As far as bounty is concerned.You can always reject my pr if not upto standard.Technical doc will be appended to readme when i push next.

**Comment by mudaly10-coder** [comment]
Auto-claim: taking this on (HAF).
Will submit a minimal PR quickly; happy to align scope/acceptance if needed.

**Comment by Somoru** [comment]
Hi,

I've just submitted PR #7441 that implements the multi-user support feature requested in this issue.

**What's Included:**
- Complete database schema migration (v234) with users, roles, user_roles, and note_shares tables
- User management service with CRUD operations and password encryption
- Role-based permission system (Admin/Editor/Reader) with granular access control
- RESTful API endpoints for user management
- Updated login flow supporting username/password while maintaining backward compatibility
- Automatic migration of existing single-user instances to admin user
- Full TypeScript type definitions

**Quality Metrics:**
✅ 948 tests passing | 17 skipped (965 total)
✅ Successful builds (server and client)
✅ Zero TypeScript errors
✅ Backward compatible with existing installations

I saw the discussion about bounties and have requested $250 for this work. Happy to discuss if needed.

The PR is ready for review - looking forward to your feedback!

**Comment by eliandoran** [comment]
@Somoru , I'll have a look at it.

**Comment by deajan** [comment]
@Somoru How does the permission inheritance system work ? Eg create a node, have sub nodes, will the sub nodes inherit node permissions ? Also, what happens with clones ? Will they inherit both "parent" nodes permissions ?

**Comment by Somoru** [comment]
@deajan

**Current Implementation:**
Right now, the permission system works at the user role level (Admin/Editor/Reader), not at the individual node level. All nodes are accessible based on the user's role.

**Permission Inheritance & Clones:**
The current PR doesn't implement per-node permissions or inheritance yet. However, I've included a `note_shares` table in the migration schema that's designed to support this functionality in the future.

**If you'd like, I can implement:**
1. Per-node permission inheritance (sub-nodes inherit parent permissions)
2. Clone permission handling (inherit from both parent branches)
3. Fine-grained sharing (share specific nodes with specific users)

These additions would significantly expand the scope of work, so I'd update the bounty request to $350 to reflect the additional implementation. Would you like me to add these features to the current PR, or would you prefer to address them separately later?

**Comment by Somoru** [comment]
@eliandoran I have made some changes according to your response and comments. Please check it out

**Comment by deajan** [comment]
@Somoru I think one of the major points of having permissions is to have them per node tree, so I'd pretty much like this, but that's not up to me to decide, but rather @eliandoran 
I'm just a feature sponsor.
But hey, don't think only bounty here. I think the 100$ more worth the full effort, but consider you'll also have some maintenance for the first releases (eg bug fixes of the code you'll provide, that you should at least follow for a couple of versions). Also, implementation should clearly be documented, in code (comments) and as a whole design in a readme/paper.

**Comment by eliandoran** [comment]
@deajan , absolutely. I was planning on setting up a policy on bounties since we need to enforce a standard of quality.

**Comment by contributor** [comment]
Please design the multi-user feature to have minimal impact on users that do not need it (probably most users?).

Currently, trilium versions 0.98-0.99.2 have serious performance issues with search. I am afraid that multi user support will degrade search even more (just think of permissions checks).

**Comment by eliandoran** [comment]
@contributor , perhaps you could offer more information about the serious performance issues with the search? Of course, in a separate issue.

**Comment by contributor** [comment]
> more information about the serious performance issues with the search? Of course, in a separate issue.

@eliandoran https://github.com/TriliumNext/Trilium/issues/6808#issuecomment-3388705542

**Comment by Somoru** [comment]
Thank you for the clarification, @deajan  — I completely understand and agree.
I’ll make sure the implementation is well-documented (both in code comments and in the design notes/README).
I’ll also stay available for follow-up maintenance or bug fixes during the first release cycles to ensure stability and smooth integration.

**Comment by shankypedia** [comment]
I just took a close look at the PR, and it seems like it was largely autopiloted.

What's Trilium's stance on AI usage? Based on the author's messages and the code, it looks like both are mostly just AI-generated filler.

**Comment by deajan** [comment]
@shankypedia I think TriliumNext uses warp.dev tool to help coding, which is exactly what @Somoru is using here.
Of course, using AI only scaffolds, and human intelligence is obviously required ^^

**Comment by shankypedia** [comment]
> [@shankypedia](https://github.com/shankypedia) I think TriliumNext uses warp.dev tool to help coding, which is exactly what [@Somoru](https://github.com/Somoru) is using here. Of course, using AI only scaffolds, and human intelligence is obviously required ^^

Using AI is fine, agreed. But it's clear when someone has no understanding of what they've implemented.

**Comment by SashankBhamidi** [comment]
@deajan @eliandoran I spent the last few days digging through Trilium's codebase to understand how multi-user could actually work. Read through all the old feature requests and realized everyone's asking for different things.

Basically there are two completely separate use cases here:

- People want isolated accounts on one server: like a family sharing a NAS. Everyone gets their own database, no note sharing needed. This is the easier one to build.

- Then there's actual collaboration: teams who need to share notes and manage permissions. That's way more complex because you need ACLs, permission inheritance, and making sure sync doesn't leak data.

I think we should support both with a config flag. Build user auth first (JWT, sessions, the basics), then add the isolated mode, then layer on collaboration features if people want it. Everything stays backward compatible.

I work in DevOps so the auth stuff is familiar territory. The hard parts are going to be permission inheritance with clones and making sure the sync system respects user boundaries.

I had AI help me map out the architecture and write documentation since there's a lot of surface area to cover. Happy to share the design docs if you want to see them.

Couple questions: 

- Does this split approach make sense to you?
- Any architectural red flags I'm missing? (Edit: You do not have access to the documents, question may be ignored).
- Do you even want this in Trilium or is single-user the intentional scope?

Also noticed another PR open for this but it looks stalled and AI-slop-y without much understanding of how sync actually works. Happy to team up if they're legit, otherwise don't want to step on toes.

What do you think?

**Comment by eliandoran** [comment]
@SashankBhamidi ,

> Basically there are two completely separate use cases here:
> People want isolated accounts on one server: like a family sharing a NAS. Everyone gets their own database, no note sharing needed. This is the easier one to build.
> Then there's actual collaboration: teams who need to share notes and manage permissions. That's way more complex because you need ACLs, permission inheritance, and making sure sync doesn't leak data.

That's right. I guess it can be implemented as multi-step phase where we first simply allow multiple users with their own database, and then implement sharing. However, we have to consider from the beginning the best database design to not block us from future expansion.

> I think we should support both with a config flag. Build user auth first (JWT, sessions, the basics), then add the isolated mode, then layer on collaboration features if people want it. Everything stays backward compatible.

I would go against having too many flags. The problem is that we would be fragmenting the application too much by having three different working modes. 

> I work in DevOps so the auth stuff is familiar territory. The hard parts are going to be permission inheritance with clones and making sure the sync system respects user boundaries.
> I had AI help me map out the architecture and write documentation since there's a lot of surface area to cover. Happy to share the design docs if you want to see them.

Sure, we can have a look. However please be aware of my comment in https://github.com/TriliumNext/Trilium/pull/7441#issuecomment-3540238170 regarding the use of LLMs and having some kind of prior experience as a Trilium developer before starting work on this huge task.

> Does this split approach make sense to you?

See previous comments.

> Any architectural red flags I'm missing? (Edit: You do not have access to the documents, question may be ignored).

> Do you even want this in Trilium or is single-user the intentional scope?

Trilium was built as a personal knowledge base which automatically implies single-user, however the request for having this has been so overwhelming that we might as well start adding support for it one way or the other.

> Also noticed another PR open for this but it looks stalled and AI-slop-y without much understanding of how sync actually works. Happy to team up if they're legit, otherwise don't want to step on toes.

For now I would keep it as a separate development. I've dismissed the original work due to the extreme use of LLMs.

**Comment by contributor** [comment]
> Trilium was built as a personal knowledge base which automatically implies single-user, however the request for having this has been so overwhelming that we might as well start adding support for it one way or the other.

Single-user is how it should stay in my opinion.

Security risks of inadvertent note leakage in multi-user setup is big deal. ACL inheritance impact is hard to reason about and hard to implement both properly and performantly. The more “properly” you implement this - the worse performance. You optimize this by cutting corners with assumptions and ACL caching - and get less “proper” implementation with edge case bugs and stale cache.

Maybe it is better to start with “multi-device” scenario?

- Single sync server - hub. Optionally, in headless mode - no web UI, except displaying publicly shared notes.
- The server can do partial sync to devices - for example, everything to desktop, and one tree branch to mobile (when we have mobile app with own db)

This is still single-user setup. But it could be extended to “family NAS” scenario to use by different people if you really don't want to run a docker container for each family member.

**Comment by SashankBhamidi** [comment]
@eliandoran, thanks for the response.

> I would go against having too many flags. The problem is that we would be fragmenting the application too much by having three different working modes.

Fair point. One clear path forward makes more sense than fragmenting the codebase.

> That's right. I guess it can be implemented as multi-step phase where we first simply allow multiple users with their own database, and then implement sharing.

This is what I've been exploring. Dug deeper into the codebase and realized the real challenge isn't authentication. It's the module-level globals: `dbConnection` in sql.ts:22, `dataKey` in protected_session.ts:5, and Becca's singleton cache. These work fine for single-user but create isolation problems when multiple users are active simultaneously.

With isolated databases per user, the auth layer routes to the correct database and each user's data stays separate. Sync works because each user's devices talk to their own database. But the globals mean you can't have true concurrent users without either spawning a process per user (approximate 100-400MB each depending on collection size, would need profiling) or refactoring all globals into CLS (more invasive).

I wrote up my findings here: https://SashankBhamidi.github.io/Trilium/

The design proposes process-per-user for now since it doesn't require touching Trilium internals. Memory overhead is real but probably acceptable for family NAS scenarios (8GB+ RAM).

> However, we have to consider from the beginning the best database design to not block us from future expansion.

Agreed. The schema I'm proposing is minimal (just a users table pointing to database paths) but doesn't block adding sharing or permissions later. Those would be separate tables that reference the user IDs.

A few things I'm less certain about:

- Is the memory overhead from process-per-user acceptable, or should I go with the CLS refactor instead?
- Where should user databases live? Thinking `./data/users/{userId}/document.db`
- Should idle user processes timeout to free memory?

This approach doesn't support note sharing between users (isolated databases by design). If collaboration is needed later, that's a different architecture requiring Becca filtering and permission systems. Not trying to tackle that yet.

One more thing: if this feature moves forward and the scope/effort makes sense to revisit the funding, I'm open to that conversation. Happy to maintain the multi-user system through a few release cycles to iron out issues.

Let me know if the approach makes sense or if there's something fundamental I'm missing.

**Comment by SashankBhamidi** [comment]
@contributor,
> Security risks of inadvertent note leakage in multi-user setup is big deal. ACL inheritance impact is hard to reason about and hard to implement both properly and performantly.

100% agree. That's exactly why the isolated database approach makes sense as a first step. No ACLs, no permission inheritance, no cache invalidation across users. Each user gets their own database file and there's literally no way for data to leak because the databases are completely separate.

> Maybe it is better to start with "multi-device" scenario?

The isolated database mode I'm proposing is essentially this. Single sync server that routes to separate databases. No shared state between users. It's multi-user in the sense of "multiple people can log into the same server" but single-user in the sense of "each person's data is completely isolated."

The full collaboration mode with note sharing and permissions is a separate (and much harder) problem. Not trying to solve that yet.

**Comment by contributor** [comment]
@SashankBhamidi 

> The isolated database mode I'm proposing is essentially this. Single sync server that routes to separate databases. No shared state between users. It's multi-user in the sense of "multiple people can log into the same server" but single-user in the sense of "each person's data is completely isolated."

Yes. But in multi-device case, it's not exactly "isolated". This is for single user who does not want to sync 1GB database to mobile app. Instead, only subset of 1GB is synced to mobile. And full 1GB is synced to desktop.

**Comment by SashankBhamidi** [comment]
@contributor 
> Yes. But in multi-device case, it's not exactly "isolated". This is for single user who does not want to sync 1GB database to mobile app. Instead, only subset of 1GB is synced to mobile. And full 1GB is synced to desktop.

That's a different use case though. You're describing one user syncing only part of their database to mobile devices. #4956 is about multiple people (family NAS) wanting separate accounts.

These are different problems:

- Your scenario: Single user, subset sync to lighter devices
- #4956 scenario: Multiple users, complete isolation between people

The isolated database approach solves the second. The first would need changes to sync logic itself, which I haven't deeply explored yet.

**Comment by contributor** [comment]
> That's a different use case though. You're describing one user syncing only part of their database to mobile devices. [#4956](https://github.com/TriliumNext/Trilium/issues/4956) is about multiple people (family NAS) wanting separate accounts.

I know. I brought it here because it is closely related. And the overall design should cover both cases. Otherwise, we may end up with "you cannot sync this part of the tree to device A because it already syncs/belongs to device B" problem.

**Comment by SashankBhamidi** [comment]
@contributor 
> I know. I brought it here because it is closely related. And the overall design should cover both cases. Otherwise, we may end up with "you cannot sync this part of the tree to device A because it already syncs/belongs to device B" problem.

Ofc both features touch the architecture.

Isolated multi-user routes requests to different databases (each has full sync). Selective device sync filters which notes sync within a single database. These are different layers and shouldn't conflict, but maintainer input on whether isolated DBs create problematic assumptions would be useful.

**Comment by eliandoran** [comment]
I've thought of isolated databases too. It works only for multi-user, no sharing, though.
Since we want to support sharing at some point, I would not sabotage that effort by choosing a design pattern that would make it more complicated for us to add it afterwards.

**Comment by SashankBhamidi** [comment]
> I've thought of isolated databases too. It works only for multi-user, no sharing, though. Since we want to support sharing at some point, I would not sabotage that effort by choosing a design pattern that would make it more complicated for us to add it afterwards.

What approach would you prefer for handling concurrent users while keeping sharing possible? The globals (`dbConnection`, `dataKey`, Becca cache) are the bottleneck.

**Comment by eliandoran** [comment]
I don't have a particular approach in mind. Most likely we'll need separate Becca instances for users in order to avoid data leaks. CLS can probably mitigate the singleton, but it would require quite a bit of work (especially on the testing side).

> > I've thought of isolated databases too. It works only for multi-user, no sharing, though. Since we want to support sharing at some point, I would not sabotage that effort by choosing a design pattern that would make it more complicated for us to add it afterwards.
> 
> What approach would you prefer for handling concurrent users while keeping sharing possible? The globals (`dbConnection`, `dataKey`, Becca cache) are the bottleneck.

**Comment by SashankBhamidi** [comment]
Understood. Sounds like this needs more architectural discussion before implementation. Happy to revisit when there's clearer direction on the sharing vs isolation trade-off.

> I don't have a particular approach in mind. Most likely we'll need separate Becca instances for users in order to avoid data leaks. CLS can probably mitigate the singleton, but it would require quite a bit of work (especially on the testing side).
> 
> > > I've thought of isolated databases too. It works only for multi-user, no sharing, though. Since we want to support sharing at some point, I would not sabotage that effort by choosing a design pattern that would make it more complicated for us to add it afterwards.
> > 
> > 
> > What approach would you prefer for handling concurrent users while keeping sharing possible? The globals (`dbConnection`, `dataKey`, Becca cache) are the bottleneck.

**Comment by deajan** [comment]
To be fair, having a Trilium instance with separate db instances is already almost doable by running with different `TRILIUM_DATA_DIR` paths and different ports behind a proxy.
That scenario would only allow to lower resource usage, which is already quite small IMO.

When I pledged for the bounty, I had real multi-user support in mind, where user A shares a note with all it's subnotres with user group B on the same instance, with read and/or write permissions.
This scenario is one of those making Trilium an option in enterprise environments.

That enterprise scenario can also fully englobe the first scenario, by adding a "no shares allowed" button, the harder part being the sync permissions, which anyway would need to be done.

I'll happily elaborate of course.

**Comment by meichthys** [comment]
I agree with @deajan 

The 'enterprise' scenario is what I was going for as well. We use Trilium in our office and just share a single instance/login. It works well but the need to be able to remove permissions from specific people is lacking. We currently 'accomplish' this via a two level (not secure) permissions structure:

Tier 1: Users with low permissions. We log these users in with the main password but don't give them the password. This keeps the user from trivially opening protected notes. (NOTE: this can probably easily be bypassed, so don't use it for highly classified material)
Tier 2: Users that know the main password and can trivially un-protect notes.

**Comment by SashankBhamidi** [comment]
Thanks for the clarity. Enterprise-grade sharing with permissions and ACL inheritance is significantly more scope than isolated databases. That requires refactoring Becca into per-user instances, implementing permission checks throughout the codebase, filtering sync operations per user.

This is the architectural surgery I was trying to avoid, but it's what the enterprise use case needs. That's beyond the scope I was targeting.

The codebase exploration was useful though - happy to answer questions if someone else picks this up.

**Comment by eitch** [comment]
When i pledged my 50$, i too was thinking of real multi-user. Same instance, same DB, multiple users with separate ACLs. As one would expect from any other multi-user system. Without looking into the code and the database structure, isn't it possible to simply map IDs of notes to a user? I.e. in a migration add a column for the user ID to each main table, in this way the user ID becomes part of each SQL query (except the joins which are probably already handled by foreign keys). This way the entry into user's notes is changed, but afterwards the rest is still standard.

But, as i said, i didn't look into the code, so i may be missing something crucial.

**Comment by deajan** [comment]
The only "hard parts" I see in real multi-user environment here are:
- The sync system where notes would be need to checked against the permissions
- The inherited permissions that could become a problem with cloned notes
- Notes ownership transfer (manual or automatic in case of owner user deletion)

I am not a trilium coder myself, but the other parts (user / group management, login flows etc) are pretty much standard to me, although they require a fair amount of time.

Using an IA, the UI for user/group management and API endpoints can be scaffolded at least, and then manually put into the shape of what's required.

There have been alot of discussions about this subject in which I've been involved. I'll make a (non AI) write up of what I think could be the right architectural way to achieve true multi-user support and then request for comments.

**Comment by rom1dep** [comment]
@deajan : if no "multiplayer" collaborative edition is considered, would the approach proposed [here](https://github.com/TriliumNext/Trilium/pull/7441#issuecomment-3432011179) work/be desirable for you?

It would allow a decoupling of concerns (and implementation phases) as follows:
- sync/federation together with permissions management: sharing a subtree with someone effectively means setting-up a new sync target on the recipient side, provided by a token bering the server URL and an authentication key
- users management: this becomes a matter of spinning trilium instances as needed, with or without a dedicated "corporate" control plane that is its own component
- central service: this becomes a matter of running users and authentication through an optional central server relaying to the different per-user instances

I suspect this delivers the same UX (or close) while buying a ton of time to rework the foundations in a truly multi-user fashion (or not!)

**Comment by itsjakobx** [comment]
https://github.com/yjs/yjs?tab=readme-ov-file

Yjs might be useful for the overreaching goal. To support real time collab. (CRDT)

It might support sqlite
<img height="300" alt="Image" src="https://github.com/user-attachments/assets/9182b08c-0c0e-454d-ad32-8b78d02674a4" />

**Comment by deajan** [comment]
@rom1dep That's already more or less doable by playing with haproxy / traefik and trilium instances with different backend ports.
I for myself run multiple trilium clients synced with multiple servers. But that's quite a mess.

**Comment by deajan** [comment]
I've re-read all exchanges about full multi-user environment for Trilium.
I've also talked about implementation with some collegues, in order to find possible gotchas and layout a design goal that won't have too much surprises later.

So here's what I came up with. Request for comments open ;)
This is a pure functional design for what to expect from a multi-user Trilium server (with "happy developpers(TM)" in mind).

# Trilium Multi user target
## Abstract

Trilium has historically been tailored for a single user environment.

The goal here is to describe what is needed to make Trilium a full blown multi-user solution which would fit for enterprises and/or family note sharing scenarios. This includes a scenario where a same Trilium instance is used by multiple users, but without any inter-user sharing capabilities.

The idea is for Trilium to become collaborative, where notes created by user A can be read / edited by user B, on the same server, based on what permissions user B has on user A's notes.

The technical implementation should derive from the functional design goals, so iterations of function implementations are designed from ground-up to fit the full multi-user scenario.

## Functional design goals

1.  Users need to be individually identified
2.  User groups need to exist to facilitate note sharing
    1.  Do we need nested groups ? Nice to have, not necessarily a hard requirement.
    2.  Users can be members of multiple groups
3.  By default, user “admin” and group “admin” must exist
4.  Permission system
    1.  All permissions can be set for users and/or groups
    2.  Permissions are:
        1.  Full: All permissions (RWX) plus permission and note ownership changes
        2.  Read: Read content of note
        3.  Write: Write content of note
        4.  Execute: As in folder traversal
            1.  This exists in sole purpose of showing that a tree exists, in order to avoid creating duplicates by users that otherwise wouldn't see existing notes
5.  Notes are owned by the user creating them
    1.  Notes owner has Full permissions on note by default
6.  Notes are only visible to the owner by default, unless other permissions apply
7.  Permission inheritance
    1.  Child notes inherit permissions from ancestors
    2.  There is no deny system, so children inherit permissions from ALL ancestors
    3.  Highest permissions always win
8.  UI
    1.  Admin ui
        1.  Simple user / group CRUD interface
            1.  Deleting user must have an additional UI to transfer note ownership or delete owned notes
        2.  Restricted to admin user and default “admin” group
        3.  Cannot remove “admin” user from admin group
    2.  Permissions UI
        1.  Permissions being set for notes, it should be done via right click > “permissions”
    3.  Permissions export
        1.  Should be able to export full notes tree with permissions for audits (optional)
9.  Synchronization
    1.  Sync clients need to authenticate as users
    2.  Sync clients only sync notes that are visible by permissions
10.  Restrictive mutli-user support
    1.  Some people on the trilium Github asked for another implementation of Trilium multi-user, where each user is isolated with his own notes, but share the same instance & database. This setup reduces resource usage and makes deployments easier.
    2.  The described “restrictive multi-user support” can be achieved by the above implementation by adding a “disable permissions” option in admin UI. In that case, every user only sees his owned notes.

## Technical implementation / requirements

*   Server has multi-user support, client is still a single user app
    *   Admin UI / REST
        *   Can only be used on server
    *   User credentials / group membership never gets synced and stays only on server
*   Automatic admin user creation with name “admin” or alike
    *   On first setup, admin user is created
    *   On existing systems, create an admin user and use current sync password as admin password
*   Automatic admin group creation, with admin user being permanent member
*   Login
    *   Login flow should always be user & password, regardless of server / client app
    *   Keep OTP if enabled
*   Note ownership transfer
    *   When deleting a user, note ownership must be transfered or notes deleted
        *   Don't allow user deletion unless notes ownership is transferred or notes deleted (enforce via API especially useful)
*   Permission inheritance
    *   On clones:
        *   Permissions from all ancestors from all clones should be merged, the highest is the winner
            * Since cloning to different subtrees would inherit other ancestor permissions, a modal which show new applied permissions would be great. Of course the modal only would appear if new permissions are inherited, so the workflow wouldn't change for single user setups
        *   Cyclic problems don't exist since they're already prohibited by cloning system
    *   Static vs dynamic permission computing:
        *   When checking permissions of a note, we need to traverse all ancestors in order to compute permissions which can quickly become expensive if a deep multi-level hierarchy exists (dynamic permissions)
        *   We might store permissions of a note in said note on creation only, and use pre-computed permissions in order to speed up permission checks (static permissions)
            *   When modifying permissions on a note, all children permissions need to be recomputed (a for loop descending to all childen)
            *   The problem here would come from clones, where the for loop would override permissions from multiple ancestor branches. Hence, static permissions would need to be stored with ancestor information, so a for loop only touches the permissions of the branch it loops over
*   Sync code
    *   Synchronization syncs all notes where Trilium client current logged-in user is owner
    *   Synchronization syncs all notes where Trilium client current logged-in user has permissions for
    *   When encountering a note with only “Execute” permission, sync note title but not content
        *   Must check if doable, depending on the database schema
*   Encryption
    *   Each user should have it's own encryption key, which could be derived from password
        *   This would require re-encryption on user password change (as with Bitwarden)
    *   Encryption keys must be synced on a per user basis, so offline Trilum clients can decrypt data

**Comment by eitch** [comment]
I love @deajan's write up! This is very much what i have in mind.

**Comment by meichthys** [comment]
> Permission inheritance
> 
>     * On clones:
>       
>       * Permissions from all ancestors from all clones should be merged, the highest is the winner
>       * Cyclic problems don't exist since they're already prohibited by cloning system

I think when cloning, it will be important to show the user which new users (if any) will have permissions on the note after the clone is created. With nested trees, it won't be immediately obvious which users have permission on the tree in which the note is being cloned, unless the clone dialog displays this.

**Comment by Aitanuqui** [comment]
@deajan, what do you have in mind for database migration, where one user wants to move from one trilium instance to a different one? The cases are the following:

1. User from single-user to a new multi user instance.
2. User from single-user to an existing multi user instance.
3. User from multi-user to a single-user instance.
4. User from multi-user to a different existing multi-user instance.

Additionally, we should take into account whether the user uses client + server or only server.

If the user has one or more client instances, I suppose that they could simply unsync and remove their database from the server. In the case of multiple client instances, they will unsync from one another unless they are synced again.

But if the user only has the server instance, what would the approach be?

For shared notes owned by the user, would unsyncing first remove all permissions for such notes from other users that had them?

**Comment by rom1dep** [comment]
@deajan excellent write-up! It's obvious (but good to state anyway) that care and attention to detail should go into indicating clearly what is shared and with whom: starting from a historically _personal_ note-taking app, we don't want to deceive users by making it too easy to share private material inadvertently.

I also regret that the "federation" aspect didn't make it to your proposal: all the people with whom _I_ currently want to collaborate already have their own trilium instance, with no meaning nor intention to ever migrate to a shared one, and no hope to ever connect as a result.

Generally speaking, I think that this proposal is mostly interesting if the implied goal is to draw to trilium a new breed of users (while doing little for the existing ones), that will use trilium in a controlled/restrictive environment (e.g. corporate setting), as a mean to an end and not by personal choice. And I am not convinced that trilium particularly shines in such deployment, while the field is already crowded with very competent, mature and corporate-friendly knowledge bases/wiki systems/CMSes.  
What makes trilium loveable and relevant today is its ability to be shaped and moulded to perfectly match one individual's mental model and creative workflows. The proposed way of doing multi-user ("using someone else's server for one job", in particular) seems largely incompatible with this.

**Comment by ascheel** [comment]
@rom1dep , a possible "resolution" to your questions surrounding federation, propagation, and shared resources could be a slightly more complex solution.  I am unsure how feasible this is, but in my head, it works relatively simply.

Trilium can be modified to connect to multiple servers simultaneously, one of which is your primary server.  Think of how Discord servers function.  Each one is in a list on the left-hand panel.  I can see this being implemented in one of two ways:

1) If you click a specific server/connection, then you get all of the contents as you would expect.  If you click another, the first is unloaded from memory and the next is loaded.  This is likely to be a bit cumbersome, so I would propose an option to only keep the last X servers loaded for performance reasons.  This would also allow the custom appCSS notes to affect only the respective server.

2) Instead of server icons or options, you simply get a new section at the bottom for "servers".  Each server expands like a Trilium item currently expands in the tree-view.  Perhaps a name such as "Server: Bob's Trilium" which would then load the root for that respective server.  This would be much more difficult to segregate how appCSS notes are handled, so my proposal would be that only appCSS notes from your home server would affect your Trilium client.  You can still view the appCSS notes from other servers, but they would be impotent, otherwise.

Both options would allow you to add or remove multiple servers without affecting the display of the others.  It would also allow easy copy/move of notes between servers.

Your "home server" would be either your local installation or a Trilium server you have access to.

**Comment by deajan** [comment]
Sorry for the long silence, I've been pretty busy with other open source projects at the time.

So summarise:

@meichthys Your assumption about clones is totally right. I have updated my writeup

@Aitanuqui Sorry but I'm not really sure what you're pointing to. Basically all Trilium instances would become multi-user, with the simple assumption that the current "admin password" becomes the admin user which would be the only user when performing an update. So basically the workflow stays the same.

> If the user has one or more client instances

What do you mean here ? As of today, a client is only synced to one server AFAIK

@rom1dep I understand the federation thing you're talking about, but that would probably become a later milestone which obviously already needs the foundations described above.

Indeed, the multi-user design is good for families and businesses. Even in the corporate world, I did not find a single tool that does the job as good as Trilium, hence the proposal.
IMHO Trilium could become the defacto tool for sharing notes in small business environments. The changes for it to become multi-user would not make it harder for current Trilium users, it merely opens the range of possibilities.

> What makes trilium loveable and relevant today is its ability to be shaped and moulded to perfectly match one individual's mental model and creative workflow

In that case, the trilium instance would be shaped into what the admin user things is good for collaboration work

@ascheel Indeed, very good idea. I already run two concurrent Trilium client instances concurrently. Perhaps allowing the same client to run multiple instances in the same window could be an improvement here.

So far, I think the 822$ as of today are far from enough for someone to really take up the work.
I could ask the company I'm working for to chip in another 500$ probably, on top of mine, but again 1322$ isn't probably worth the effort.
We'd probably need a developper who does it "for the fun" as some of us already do for our own open source projects.

**Comment by meichthys** [comment]
> IMHO Trilium could become the defacto tool for sharing notes in small business environments.

It is for my company already, but this feature would significantly improve the situation. 👍

**Comment by zhanglin2603** [comment]
Hey, I went over TriliumNext/Trilium#4956 and can take this on (listed bounty: $822).
- My read: this is primarily a core implementation task tied to "Milestone: Multi-user support".
- I will focus on these issue-specific points first: milestone, multi-user, issuehunt.
- First I will reproduce and confirm expected behavior for the core implementation path.
- Then I will implement the fix with targeted tests and keep scope tight (high confidence).
- I will share a working PR update fast (24h for scoped implementation plan, 48h+ for full patch and validation) with verification notes.
If this approach works for you, I can start now.

**Comment by deajan** [comment]
@zhanglin2603 As long as it won't be an AI contribution you're welcome. This feature has deep implementation implications an AI won't be able to understand.

**Comment by eliandoran** [comment]
Hi, @zhanglin2603 .

FYI, in order to benefit of the bounty, your architecture and work **must not** be LLM-generated. I would say the maximum threshold for using LLMs for a feature like this is ~10%.

Before starting any development work, it's best to share a plan and the architecture with a focus on the hard problems that were discussed here.

Most importantly, I would expect the developer to have a fairly good knowledge of Trilium Notes as a user.

**Comment by zhanglin2603** [comment]
Thanks for the clarification. Understood on the implementation depth concern; I’ll step back from this issue.

**Comment by SashankBhamidi** [comment]
@eliandoran @deajan

Been a few months. I kept coming back to this problem for every comment posted (I subscribed to the thread).

Went back through the codebase. CLS already handles per-request state, so I built the architecture around extending it for per-user Becca, encryption keys, and identity. Shared database stays intact, no process-per-user overhead, no config flags.

Rewrote the design doc: https://SashankBhamidi.github.io/Trilium/

Few questions at the end. Want to get those sorted before writing code.

What do you think?

**Comment by deajan** [comment]
@SashankBhamidi I'll read it as soon as I get some quality off time and report back.

**Comment by mingefei-cloud** [comment]
## IssueHunt Bounty Submission

**Task:** Milestone: Multi-user support
**Repository:** TriliumNext/Trilium
**Issue:** #4956
**IssueHunt Link:** Not specified

**Completion Status:** ✅ Analysis and execution plan ready
**Completed:** 2026-03-01T07:11:02.220Z
**GitHub User:** mingefei-cloud
**IssueHunt Account:** Connected via GitHub

**Bounty Details:**
- Estimated bounty: $822
- Skill match:  (0%)
- Priority score: 339

**Execution Plan:**
- **Platform:** IssueHunt
- **Type:** issuehunt_bounty
- **Estimated Time:** 2-6 hours
- **Approach:** Complete implementation with IssueHunt integration
- **Steps:** Analyze IssueHunt bounty requirements → Implement solution according to specifications → Test implementation thoroughly → Provide proof of completion → Submit via IssueHunt platform if required
- **Deliverables:** Code implementation, Documentation, Test results, IssueHunt submission proof

**Autonomous Agent Details:**
- Agent: OpenClaw IssueHunt Monitor
- GitHub: mingefei-cloud
- Execution timestamp: 3/1/2026, 3:11:02 PM

**Next Steps:**
I will proceed with the implementation according to the IssueHunt bounty requirements and provide updates upon completion.

**Comment by eliandoran** [comment]
@mingefei-cloud , no AI-driven activity, please. Even the skill match is zero, according to your comment.

**Comment by TheBarnyOfBarnim** [comment]
@eliandoran
I feel like that @mingefei-cloud may be a bot...

**Comment by Asobu01** [comment]
I’d like to claim `TriliumNext/Trilium#4956` and can start immediately. I’ll share a short execution plan first, then submit a review-ready PR within 24 hours. I reduce back-and-forth by posting checkpoints early and attaching clear validation evidence with the PR. I am currently tracking 22 active bounty thread(s) with timely updates. (@Asobu01)

**Comment by eliandoran** [comment]
@TheBarnyOfBarnim , yup, most likely. Apparently bounties attract a lot of unwanted attention.

**Comment by eliandoran** [comment]
@Asobu01 , given you have no prior experience with Trilium and your GitHub account is new and AI-oriented, the answer is no.

**Comment by SashankBhamidi** [comment]
> [@SashankBhamidi](https://github.com/SashankBhamidi) I'll read it as soon as I get some quality off time and report back.

@deajan Did you get a chance to take a look?

**Comment by Ntropic** [comment]
For true multi user support, wouldn't it make sense to first add support for other, concurrent capable databases, see [5347](https://github.com/TriliumNext/Trilium/issues/5347?utm_source=chatgpt.com)

**Comment by zekageri** [comment]
I recently switched from Joplin because I was looking for a nicer UI, and I found Trilium. It is a fantastic piece of work, and I really love the app.

In Joplin, my wife and I used to share notes with each other. We both really like Trilium, but the fact that it only supports one user and that multiple instances cannot share notes in any way is a bit disappointing for our use case.

It would be really nice to have some kind of export/import feature for individual notes or note trees. For example, I could export an entire note and send it to my wife, and she could import it into her own Trilium instance.

This would not need to be full multi-user collaboration, but even a simple way to share notes between separate instances would be very useful.

**Comment by meichthys** [comment]
@zekageri There is already export/import support. simply right click on any note in the tree and choose "Export" or "Import into note". 
You can also click the `...` note menu and choose "Export note" or "Import files":

<img width="269" height="469" alt="Image" src="https://github.com/user-attachments/assets/0664a6ee-7e40-4a6f-b3e8-bcea74c6fcf5" />

**Comment by zekageri** [comment]
Oh, I somehow missed that. So many features. Thank you.

I think I will give a shot to the multi container way now.

**Comment by xcq20100224** [comment]
Hi maintainers, I’m interested in taking over this bounty and moving it forward from the previously closed draft PR (#7441).

To reduce risk and make review manageable, would you accept a phased delivery approach?

Proposed Phase 1 (review target):
- Introduce a minimal multi-user foundation (account/session model and isolation boundaries)
- Keep compatibility with current single-user behavior by default
- Add initial tests/docs for the new flow and migration notes

Then follow-up phases can cover permissions/ACL hardening, admin UX, and broader edge cases.

If this scope is acceptable, I can start with a clean PR for Phase 1 and iterate quickly on maintainer feedback.
Please confirm whether this phased approach is acceptable for issue #4956 bounty progression.
Thanks!

**Comment by deajan** [comment]
@xcq20100224 Will this be an AI contribution ? What is your skill level on typescript and general SQL knowledge ?
This bounty is critical and cannot be an AI contrib.

**Comment by xcq20100224** [comment]
Thanks @deajan for the quick response.

This would be a human-led contribution. I may use AI for drafting/boilerplate productivity, but architecture decisions, implementation, testing, and review will be done by me.

Skill-wise, I’m comfortable with TypeScript and SQL for backend-oriented work (schema changes, migrations, indexing/performance basics, and integration debugging).

To reduce risk, I’d start with a small reviewable Phase 1 PR (minimal account/session foundation + isolation boundaries + compatibility default + tests/docs), then iterate based on maintainer feedback.

If maintainers agree with this approach, I can open the Phase 1 PR next.

**Comment by deajan** [comment]
It's up to @eliandoran to set boundaries here, but IMO PR [7441](https://github.com/TriliumNext/Trilium/pull/7441) was AI made only, hence it wouldn't be a good starting point.

Please read the following lecture before anything:
- https://github.com/TriliumNext/Trilium/issues/4956#issuecomment-3572299547
- https://sashankbhamidi.github.io/Trilium/

If you feel comformtable with a non AI implementation, let us know.

**Comment by deajan** [comment]
@SashankBhamidi I'm really sorry, I totally forgot to give feedback on your writeup I read months ago. So this sunday, I decided to read it again, and summon my earlier thoughts.

### On the 6 things that break:
The database connection: Why would you need a context aware database conenction ? If permissions are implemented by the software, it's up to the software to have `WHERE user_id = '<id>'` statements or alike.
On the server side, this would not be a problem. On the client side, since sync would only pick up notes which have user permissions, I don't see a problem neither.

### Permission inheritance with clones:
Indeed performance of permission recomputation is O(depth * branches). Good point.

### Becca memory
Having la LRU becca instance is a really good idea, nice catch.

### Selective per-device sync
While doing the permissions check in the sync logic, it would be easy to add a "device permission" which would restrict some notes to some devices. I don't really see the necessity of such functionnality right now, just my two cents as why it would be connected to current plans.

### Nested groups
As @meichthys suggested, it would make less obvious which user sees some notes, and it's not a hard requirement unless someday Trilium becomes a fully fleged enterprise setup.

### User table migration
I would suggest that user table also contains optional email address, so future SSO usages would be covered.

@SashankBhamidi Again, really sorry for the (very) late feedback.

**Comment by SashankBhamidi** [comment]
@deajan No need to apologize. Going through your points:

> The database connection: Why would you need a context aware database conenction ? If permissions are implemented by the software, it's up to the software to have WHERE user_id = '<id>' statements or alike.

Yes it is right that permission filtering doesn't need a context-aware connection, the `WHERE` clauses handle that. The reason I put it in the list is the prepared-statement cache that lives on the connection. So with one shared connection, two users hitting the same query paths churn each other's cache. And with per-user instances, each user's hot queries stay warm. So it is performance and not correctness, and the design doc should say that cleary.

Also worth knowing: this one's already most of the way there! The SQL layer was refactored into an instantiable `SqlService` class (`packages/trilium-core/src/services/sql/sql.ts:17`) with a per-instance statement cache. So the bullet shifts from "we need to refactor this" to "instantiate it per user when the time comes." I'll downgrade it in the next pass.

> Permission inheritance with clones: Indeed performance of permission recomputation is O(depth * branches). Good point.

Yeah the static-caching approach gets us O(1) reads but pays on writes. Most permission changes touch one subtree, so the amortized cost is fine in practice. The failure mode is deep trees with frequent permission shuffles. It will be worth profiling against a real database before committing to an invalidation strategy.

> Becca memory: Having la LRU becca instance is a really good idea, nice catch.

The eviction window is the main lever. Too aggressive and we re-hydrate from disk constantly, too lax and memory creeps up over time :)

30 min was my starting guess but the right value will depend on the deployment, so I'd make it configurable.

> Selective per-device sync: While doing the permissions check in the sync logic, it would be easy to add a "device permission" which would restrict some notes to some devices.

Yes this one falls out almost for free once permission-filtered sync exists. A `device_permissions` table, or even just a flag on the existing permission rows is enough. Not building it day one, but the schema should stay shaped so it's possible to add later without a migration headache.

> Nested groups: As @meichthys suggested, it would make less obvious which user sees some notes, and it's not a hard requirement unless someday Trilium becomes a fully fleged enterprise setup.

I agree, leaving it out of Phase 1. If it ever becomes a hard requirement, a `parentGroupId` column is a one-line addition and the inheritance code is simpler than note-permission inheritance (no clones to worry about), so it's a small lift to add later.

> User table migration: I would suggest that user table also contains optional email address, so future SSO usages would be covered.

Good idea. Also it is worth me flagging that `user_data` already has an `email` column today (OpenID uses it), so the new `users` table should match that shape with an optional `email TEXT` plus an index. Keeping it optional so single-user setups don't have to fill anything in, and so it's ready for SSO when that path gets used. And probably also worth thinking about whether `users` just absorbs `user_data` entirely on the migration side, since otherwise we'd have two places storing the same thing.

One more thing I noticed while re-validating against current upstream: I owe the doc a citation refresh. The codebase moved into `packages/trilium-core/` and most of my file:line references are stale now. The architecture argument still holds, the line numbers don't. I'll do that in the same pass as the changes above.

**Comment by eliandoran** [comment]
I've decided to unilaterally refund all the bounties that we had on IssueHunt.

When I first set up the bounties, the concept made perfect sense since people could easily "vote with their wallets" their favorite features which would also be an incentive mostly to the dev team to implement said features. Now with the AI bubble, the barrier to creating PRs is almost zero which means that we have had a significant influx of contributions meant only to receive said bounty. This is not something we encourage because it puts an unequal burden on us as reviewers, since the cost of a review is higher than the development cost (when AI is used).

When refunding I got this message from IssueHunt saying that the Stripe fee will not be refunded. Apologies for the lost funds, but perhaps it's for the best. I would prefer to encourage a tip culture instead; when a feature lands that you are happy with, simply use the normal donation channels. This also unblocks your funds since sometimes it can take a lot of time to implement a feature.

If you have any issues or concerns, let me know.

**Comment by eitch** [comment]
I'm fine with this. I would have also simply donated a few dollars, but thought it would be nice to put the money to ma most burning feature request.

**Comment by deajan** [comment]
Understandable.
This feature really would make my life a lot easier.
Hopefully it will get some traction.

**Comment by SashankBhamidi** [comment]
@deajan working on it. Pushed an update to the doc a few days back: [SashankBhamidi.github.io/Trilium](https://SashankBhamidi.github.io/Trilium/). Citations now point at upstream SHA `35bc738` (May 21, 2026) so they hold up over time.

A few things came out of your last review. The SQL bullet is downgraded, you were right that filtering doesn't need a context-aware connection. The `users` table gets an `email` column, and there's an open question about whether the new table just absorbs `user_data`. I also added a section explaining why `entity_changes.userId` has to land in Phase 1 even though sync filtering doesn't turn on until Phase 3, that one was missing.

@eliandoran @deajan would love to know what you both think when you've had a chance to read it. If the architecture lands, I'll start on Phase 1.

**Comment by Leonardo-Costa** [comment]
Hi, wondering if there has been any update on this feature. Has this joined any planned roadmap?

**Comment by eliandoran** [comment]
@SashankBhamidi , feel free to do a PoC on the phase 1 to see how it goes.

Regarding your open questions:

1. `user_data` is quite a strange table because it's coded in such a way that it supports only one user anyway. It needs to go away. A one-time migration should not be a problem, only the OAuth secrets need to be carried over to the new table. Note that we've recently made changes to the OAuth mechanism to actually bind to a particular user, as we missed the feature in the past.
2. Honestly, I think attribute-level permissions would cause too much complexity and make debugging a nightmare. Attributes are part of the note in the same way the content is a part of a note.
3. In general Becca can cause issues with memory when there are many users involved. I don't see the search index being a problem until much later in the development of multi-user.

@Leonardo-Costa , this is **not** yet planned on a roadmap.

**Comment by philippl-onz** [comment]
as someone who wants to use trilium purely in browsers (running on my home kubernetes cluster) and sharing notes with my family i would love to see this feature. 

Also basically ALL notes apps are single user (local style) only so this feature would be an awesome USP.

**Comment by nealhamiltonjr** [comment]
I'm reviewing docmost and trilium. I like trilium more for the structure, encryption and web scraping and a few more. It feels more robust for building a higher quality documentation center. It's everything a wiki should be except the multi user support. I understand it wasn't originally designed for that, just saying it's so close to being one of the best ones out there despite its beginnings. Maybe we could get the tiptap editor, I find it better for code related stuff. Can Trilium have both editors and let the user choose? Would be a great wiki for tech shops. And the tiptap one has inline highlighting for notes and being able to move paragraphs around..both are god sends. 

Thank you for all the hard work you devs put into this.

**Comment by TheFreeman** [comment]
I have opened a focused design proposal for owner-based personal branch sharing:

https://github.com/orgs/TriliumNext/discussions/10687

The proposal builds on the ongoing multi-user work and defines:

- personal note roots for every user;
- explicit and persistent ownership;
- inherited read/write branch grants;
- owner-preserving creation of descendant notes;
- safe handling of branches, clones, relations, and private paths;
- service-layer authorization;
- Becca, search, cache, ETAPI, WebSocket, and synchronization behavior;
- restrictions for protected and executable notes;
- migration and incremental implementation phases.

The goal is deliberately not to turn Trilium into a team wiki, but to preserve the model:

> “This is my Trilium. I may allow someone to read or help edit this one branch.”

**Comment by nealhamiltonjr** [comment]
We need a modern editor like tiptap, the current one is god awful.
